<!doctype html>
<html lang="en">
    <head>
        <meta http-equiv="content-type" content="text/html; charset=utf-8" />
        <meta name="viewport" content="width=device-width, initial-scale=1.0" />
        <link rel="icon" type="image/svg+xml" href="/assets/images/small_logo.svg" />
        <title>GP Map: FAQ</title>
        <script type="module" src="/main.js"></script>
        <script async src="https://www.googletagmanager.com/gtag/js?id=G-JRYWRJTDBP"></script>
        <script>
            window.dataLayer = window.dataLayer || [];
            function gtag() {
                dataLayer.push(arguments);
            }
            gtag("js", new Date());

            gtag("config", "G-JRYWRJTDBP");
        </script>
    </head>

    <body>
        <navigation-bar></navigation-bar>
        <div class="content columns mt-2 is-centered">
            <div x-data="faq" class="column is-11">
                <h1>FAQ</h1>
                <h2 id="r-package-view">R Package</h2>

                <p>
                    <a href="https://github.com/MRCIEU/gpmapr">The R package</a> is used to download the data from the
                    API and perform the colocalization and rare variant analysis.
                </p>
                <ul>
                    <li>
                        <b>Installation and Usage</b><br />
                        <code
                            >devtools::install_github('MRCIEU/gpmapr')<br />library(gpmapr)<br />search_gpmapr('haemoglobin')</code
                        >
                    </li>
                    <li>
                        <b>Usage</b>: There are a series of vignettes available.
                        <a href="https://mrcieu.r-universe.dev/articles/gpmapr/gpmapr.html"
                            >Please start with the intorudctory vignette.</a
                        >
                    </li>
                </ul>
                <h2 id="gwas-upload-pipeline-view">GWAS Upload Pipeline</h2>
                <img :src="gwasUploadPipeline" alt="GWAS Upload Pipeline" />
                <p>
                    The GWAS upload pipeline is a process that allows you to upload a GWAS and perform the
                    colocalization and rare variant analysis. It uses the same data processing pipeline used to created
                    this resource, with some caveats. There are a series of steps that are taken to process the data,
                    some of which will remove the data and make it look potentially inconsistent.
                </p>
                <ul>
                    <li>
                        <b>Filtered list of comparisons:</b> Due to server constraints, only studies that have a minimum
                        p-value of 1e-6 for that ld block will be compared with your GWAS.
                    </li>
                    <li>
                        <b>No GWAS QC step:</b> Due to server constraints, DENTIST (which is used in the pipeline) is
                        not run on any GWAS Uploads.
                    </li>
                    <li>
                        <b>Sparesly Populated Studies Not Supported:</b> Only studies with a minimum of 150 samples in a
                        significant ld block will be processed.
                    </li>
                    <li>
                        <b>Rare Variant Analysis Not Supported:</b> Only samples with a MAF of 0.01 or greater will be
                        processed.
                    </li>
                    <li>
                        <b>Conditional Imputation:</b> Imputation is only performed if the correlation between origina
                        and imputed SNPs is greater than 0.7.
                    </li>
                    <li>
                        <b>Conditional Finemapping:</b> If finemapping only finds a single credible set or does not
                        converge, then the LBF values are merely calculated from the original summary statistics.
                    </li>
                    <li>
                        <b>Missing LBF Values:</b> There are a series of finemapping filtering steps that find and
                        remove over inflated LBF values, these are removed from the analysis. Hence you may see some
                        missing LBF values in the results.
                    </li>
                </ul>
                <h2 id="graph-options-view">Graph Options</h2>
                <ul>
                    <li><b>Study P-value:</b> The p-value threshold for the traits to be displayed in the legend.</li>
                    <li><b>Include Trans Markers:</b> Whether to include trans markers in the graph.</li>
                    <li>
                        <b>Trait Types:</b> The type of trait to be displayed in the graph. 'Molecular Only' will still
                        include the phenotype in question on the phenotype view, 'Phenotype Only' will not include
                        molecular traits.
                    </li>
                    <li>
                        <b>Trait Categories:</b> The category of trait to be displayed in the graph. If no categories
                        are selected, all traits will be displayed.
                    </li>
                </ul>
                <h2 id="trait-view">Trait View</h2>
                Displays colocalised results of the study in question, and shows all studies which colocalise with it,
                overlayed on top of the the Manhattan plot of the phenotype. Also displays significant rare and
                non-colocalising results. To compare 2 specific traits, please use the 'Filter Results By' dropdown
                <ul>
                    <li>
                        <b>Colocalised Results:</b> Displays colocalised results of phenotype in question, and shows all
                        studies which colocalise with it
                    </li>
                    <li>
                        <b>Rare association Results:</b> These are not colocalization groups, but single SNP
                        associations that both show significant association with the phenotype in question.
                    </li>
                    <li>
                        <b>Circle size:</b> The size of the circle is proportional to the number of traits in the
                        colocalisation group. The larger the circle, the more traits are in the colocalisation group.
                    </li>
                    <li>
                        <b>Result Filtering:</b> To compare 2 specific traits, please use the 'Filter Results By'
                        dropdown, this will filter the results to only show the traits that are selected.
                    </li>
                </ul>
                <h2 id="variant-view">Variant View</h2>
                SNP view displays the colocalisation results for a single SNP. Each circle represents a trait that is in
                the colocalisation group.
                <ul>
                    <li>
                        <b>How the SNP is chosen:</b> The SNP is chosen as the cumulative sum of the log bayes factor,
                        which is returned by <a href="https://github.com/stephenslab/susieR">susie</a>. Every trait in
                        the colocalisation group is included in the cumulative sum, and the maximum is taken.
                    </li>
                    <li>
                        <b>Node size:</b> Each node is sized by the p-value of the SNP for that specific trait, the
                        smaller the p-value, the larger the node.
                    </li>
                    <li>
                        <b>Links:</b> Each link represents a colocalization pair result, as returned by
                        <a href="https://chr1swallace.github.io/coloc/">coloc</a>.
                    </li>
                    <li>
                        <b>Link Strength:</b> The strength of the link is displayed as the H4 value, which is a measure
                        of the strength of the colocalization pair result. A significant link (H4 > 0.8) is displayed in
                        blue, a weak link (0.8 > H4 > 0.5) is displayed in orange.
                    </li>
                    <li>
                        <b>Group Connectedness:</b> The connectedness of the colocalisation group is calculated as the
                        sum of the H4 values of the coloc pairs, divided by the total number of coloc pairs in the
                        gorup. A higher connectedness means that the colocalisation group is more strongly connected to
                        other colocalisation groups.
                    </li>
                    <li><b>Trait Type:</b> The type of trait is displayed in the legend.</li>
                    <li>
                        <b>Common vs Rare:</b> The common groups are visualised in the graph, but the rare results are
                        not, they are included in the results table and forest plot
                    </li>
                    <li>
                        <b>VEP annotation:</b> The data displayed under VEP annotation is data related to the SNP, as
                        provided by the
                        <a href="https://www.ensembl.org/info/docs/tools/vep/index.html"
                            >Ensembl Variant Effect Predictor</a
                        >
                    </li>
                </ul>
                <h2 id="gene-region-view">Gene and Region View</h2>
                Displays the colocalisation results for a gene or region. Each circle represents a result that has a
                study marked with that gene. Results may not align with the exonic region of the gene, as some studies
                may have QTLs which are in the regulatory region of the gene.
                <ul>
                    <li>
                        <b>Colocalised Results:</b> Displays colocalised results of phenotype in question, and shows all
                        studies which colocalise with it
                    </li>
                    <li>
                        <b>Rare association Results:</b> These are not colocalization groups, but single SNP
                        associations that both show significant association with the phenotype in question.
                    </li>
                    <li>
                        <b>Circle size:</b> The size of the circle is proportional to the number of traits in the
                        colocalisation group. The larger the circle, the more traits are in the colocalisation group.
                    </li>
                    <li>
                        <b>Surrounding Genes:</b> Below the graph, the surrounding genes are displayed. These are genes
                        that are in the same region as the gene in question, and have a study marked with them.
                    </li>
                    <li>
                        <b>Result Filtering:</b> You can filter by another surrounding gene by clicking on the gene in
                        question. To compare 2 specific traits, please use the 'Filter Results By' dropdown, this will
                        filter the results to only show the traits that are selected.
                    </li>
                    <li>
                        <b>Pleiotropy Scores:</b> There are two pleiotropy scores displayed, the first is the number of
                        distinct trait categories and the second is the number of distinct protein coding genes that the
                        trait is associated with. They are calculated by counting the number of distinct trait
                        categories and protein coding genes that the gene is associated with in the colocalisation
                        results. Rare variant results are not included in the calculation.
                    </li>
                </ul>
            </div>
        </div>
    </body>
</html>
